### ISSUE TITLE
Flaky-test CI bug: weird long delays on JS

### ISSUE THREAD

**Issue by noahgibbs** [scarpe-team/scarpe#272]
This is a funky one. It's not sending heartbeats, so I opened a bug on that. But it's not clear it should be yet -- the early stuff, navigate and initial draw, go *incredibly* slowly.

Init takes well over 2 seconds (note jump from 514 to 2888), which is fine. But there's that big jump from the init event going out (2889ms) until the timeout (7586ms). Doesn't even request DOM replacement until 7584. So it's like it's just waiting for almost 5 seconds before even requesting the DOM replacement. Something is funky there, just not sure what.

event log exerpt:
```
[514] DEBUG DisplayService: Subscribe to event: "destroy" T: 2320
[2888] DEBUG DisplayService: Subscribe to event: "init" T: nil
(...)
[2889] DEBUG DisplayService: Dispatch event: "run" T: nil A: [] KW: {}
[2889] DEBUG WV::ControlInterface: CTL event :init [] {}
[7586] WARN  WV::ControlInterface: die_after - timed out after 4.697416 (threshold: 3.0)
[7587] DEBUG WV::ControlInterface: Writing results file "/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/scarpe_test_results.json20230630-1475-8r1v1c" to disk!
[7668] DEBUG WV::ControlInterface: CTL event :shutdown [] {}
[7671] DEBUG WV::ControlInterface: CTL event :heartbeat [] {}
```

main log excerpt:

```
[11] INFO  LoggedScarpeTest: Test: TestControlInterface#test_assert_widget
[516] DEBUG WV::WebWrangler: Creating WebWrangler...
[2889] DEBUG WV::WebWrangler: Run...
[7584] DEBUG WV::WebWrangler::DOMWrangler: Requesting DOM replacement...
[7584] DEBUG WV::WebWrangler::DOMWrangler: Requesting redraw with 1 waiting changes and no waiting promise - need to schedule something!
[7584] DEBUG WV::WebWrangler::DOMWrangler: Requesting redraw with 1 waiting changes - scheduling a new redraw for them!
[7584] DEBUG WV::WebWrangler: Scheduled JS: (0)
(function() {
  var code_string = "document.getElementById('wrapper-wvroot').innerHTML = `<div style=\"height:100%\"><p id=\"2500\" style=\"font-size:12px\">Hello World</p></div>`; true";
  try {
    result = eval(code_string);
    scarpeAsyncEvalResult("success", 0, result);
  } catch(error) {
    scarpeAsyncEvalResult("error", 0, error.message);
  }
})();

[7668] DEBUG WV::WebWrangler: Destroying WebWrangler...
[7673] DEBUG WV::WebWrangler: Got JS value: success / 0 / true
[7674] DEBUG WV::WebWrangler::DOMWrangler: Fulfilled redraw with no waiting chan
ges - marking us as up to date!
[7674] DEBUG WV::WebWrangler::DOMWrangler: Redraw is now fully up-to-date
[8831] INFO  TestHelper: JSON assertion data: [false, "Timed out!", {"die_after"
=>{"t_start"=>"2023-06-30 06:15:24 +0000", "threshold"=>3.0, "passed"=>4.697511}
}]
```

[test failure logs.zip](https://github.com/scarpe-team/scarpe/files/11926491/test.failure.logs.zip)

**Comment by noahgibbs** [comment]
In the WV API log, there's another oddity...

Looking below, it *is* doing things between the 2880ms time and the 7500ms time. It's running JS against Webview. It's just taking forever. Calls set_size at 3009, then over 2 seconds later finally calls navigate at 5537. Then navigate is instant, and it tries to patch console at 5538. Doesn't even *start* trying to set up the DOM until 7584, by which time it's long timed out.

```
[2934] INFO  WebviewAPI: Method: set_title Args: ["Scarpe!"] KWargs: {} Block: n Return: nil
[3009] INFO  WebviewAPI: Method: set_size Args: [480, 420, 0] KWargs: {} Block: n Return: nil
[5537] INFO  WebviewAPI: Method: navigate Args: ["data:text/html, %3Chtml%3E%0A++%3Chead+id%3D%27head-wvroot%27%3E%0A++++%3Cstyle+id%3D%27style-wvroot%27%3E%0A++++++%2F%2A%2A+Style+resets+%2A%2A%2F%0A++++++body+%7B%0A++++++++font-family%3A+arial%2C+Helvetica%2C+sans-serif%3B%0A++++++++margin%3A+0%3B%0A++++++++height%3A+100%25%3B%0A++++++++overflow%3A+hidden%3B%0A++++++%7D%0A++++++p+%7B%0A++++++++margin%3A+0%3B%0A++++++%7D%0A++++%3C%2Fstyle%3E%0A++%3C%2Fhead%3E%0A++%3Cbody+id%3D%27body-wvroot%27%3E%0A++++%3Cdiv+id%3D%27wrapper-wvroot%27%3E%3C%2Fdiv%3E%0A++%3C%2Fbody%3E%0A%3C%2Fhtml%3E%0A"] KWargs: {} Block: n Return: nil
[5538] INFO  WebviewAPI: Method: eval Args: ["function patchConsole(fn) {\n  const original = console[fn];\n  console[fn] = function(...args) {\n    original(...args);\n    puts(...args);\n  }\n};\npatchConsole('log');\npatchConsole('info');\npatchConsole('error');\npatchConsole('warn');\n"] KWargs: {} Block: n Return: nil
[7584] INFO  WebviewAPI: Method: eval Args: ["(function() {\n  var code_string = \"document.getElementById('wrapper-wvroot').innerHTML = `<div style=\\\"height:100%\\\"><p id=\\\"2500\\\" style=\\\"font-size:12px\\\">Hello World</p></div>`; true\";\n  try {\n    result = eval(code_string);\n    scarpeAsyncEvalResult(\"success\", 0, result);\n  } catch(error) {\n    scarpeAsyncEvalResult(\"error\", 0, error.message);\n  }\n})();\n"] KWargs: {} Block: n Return: nil
[7671] INFO  WebviewAPI: Method: terminate Args: [] KWargs: {} Block: n Return: nil
```

**Comment by noahgibbs** [comment]
The test itself is really simple, not doing anything funny:

```
  def test_assert_widget
    run_test_scarpe_code(<<-'SCARPE_APP', test_code: <<-'TEST_CODE')
      Shoes.app do
        para "Hello World"
      end
    SCARPE_APP
      on_event(:next_redraw) do
        para = find_wv_widgets(Scarpe::WebviewPara)[0]
        assert para
        return_when_assertions_done
      end
    TEST_CODE
  end
```

**Comment by noahgibbs** [comment]
One possibility is just to assume Webview can take a fair bit of time to load on CI. If I just adjusted thresholds to have a minimum of 10 seconds for each test, that *might* make it basically never flake.

Or not.

**Comment by noahgibbs** [comment]
Should be fixed by #289.
